sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -61,8 +62,18 @@ object ConstantFolding extends Rule[LogicalPlan] {
       case Size(c: CreateMap, _) if c.children.forall(hasNoSideEffect) =>
         Literal(c.children.length / 2)
 
-      // Fold expressions that are foldable.
-      case e if e.foldable => Literal.create(e.eval(EmptyRow), e.dataType)
+      // Fold expressions that are foldable. To avoid repeatedly evaluating expressions which throw
+      // exceptions, we also check whether the expression's children are all literals.
+      case e if e.childrenToEvaluate.forall(_.isInstanceOf[Literal]) && e.foldable =>
+        val constant = try {
+          Some(e.eval(EmptyRow))
```

Comment:
I think we should do this in `Expression.foldable`. Basically if overflowing can happen, they are not foldable.

Reply:
hmm, then it's a big perf regression for ANSI mode, as all math operations can't do constant folding.

Reply:
We can add it only into the expressions that can do constant folding I guess.

Reply:
According to @cloud-fan 's advice, I refactor this patch by using `RaiseError` as constant. Now we don't need to bother with the foldable expressions which can throw exceptions, cause they are still foldable and will be folded to `RaiseError`s.